share/revng/test/tests/abi/gather_symbols.py

```diff
@@ -0,0 +1,398 @@
+        data,
+    )
+    if match:
+        return match[1].decode("utf-8")  # address
+    else:
+        sys.exit(1)
+
+
+def find_section(name, data):
+    name_bytes = bytes(name, "utf-8")
+    match = re.search(
+        b"\n\s+\d+\s+"  # index
+        + name_bytes  # name
+        + b"\s+([a-fA-F0-9]+)\s+"  # size
+        + b"([a-fA-F0-9]+)\s+"  # vma
+        + b"([a-fA-F0-9]+)\s+"  # lma
+        + b"([a-fA-F0-9]+)\s+"  # offset within the file
+        + b"\d+\*\*\d+\s*",  # (uncaptured) alignment
+        data,
+    )
+    if match:
+        return (
+            match[1].decode("utf-8"),  # size
+            match[2].decode("utf-8"),  # vma
+            match[4].decode("utf-8"),  # offset
+        )
+    else:
+        sys.exit(2)
+
+
+def find_callsites(name, data):
+    result = []
+
+    name_bytes = bytes(name, "utf-8")
+    for match in re.finditer(
+        b"\n\s+0*([a-fA-F0-9]+):"  # address (without leading 0s)
+        + b"[^<\n]*<"  # any number of non-`<` characters followed by `<`
+        + name_bytes  # name
+        + b">\n"  # new line
+        + b"\s+0*([a-fA-F0-9]+)"  # address of the next instruction
+        + b"[^\n]*\n"  # remainder of the line
+        + b"\s+0*([a-fA-F0-9]+)",  # address of one more instruction
+        data,
+    ):
+        call_instruction_address = match[1].decode("utf-8")
+        next_instruction_address = match[2].decode("utf-8")
+        one_more_instruction_address = match[3].decode("utf-8")
+
+        call_size = int(next_instruction_address, 16) - int(call_instruction_address, 16)
+        next_size = int(one_more_instruction_address, 16) - int(next_instruction_address, 16)
+        result.append(
+            (call_instruction_address, str(call_size), next_instruction_address, str(next_size))
+        )
+
+    return result
+
+
+def find_first_instruction_size(address, data):
+    address_bytes = bytes(address, "utf-8")
+    match = re.search(
+        b"\n\s+0*"  # leading spaces and 0s
+        + address_bytes
+        + b"\s+<\w+>:\n"  # function name
+        + b"\s+0*"  # leading spaces and 0s
+        + address_bytes
+        + b":[^\n]*\n"  # remainder of the line
+        + b"\s+0*([a-fA-F0-9]+):",  # next address (without leading 0s)
+        data,
+    )
+    if match:
+        return int(match[1].decode("utf-8"), 16) - int(address, 16)
+    else:
+        sys.exit(3)
+
+
+@dataclass
+class Section:
+    name: str = ""
+    size: str = ""
+    vma: str = ""
+    file_offset: str = ""
+
+
+@dataclass
+class MemoryLocation:
+    name: str = ""
+    address: str = ""
+
+
+@dataclass
+class CallSite:
+    before: str = ""
+    before_size: int = 0
+    after: str = ""
+    after_size: int = 0
+
+
+class Function(MemoryLocation):
+    def __init__(self, name: str, address: str, entry_point: str):
+        self.name = name
+        self.address = address
+        self.entry_point = entry_point
+        self.callsites: list[CallSite] = []
+
+
+class Gathered:
```

Comment:
Gathered?

Reply:
Not much thought into the name. Since the script is `gathering` stuff, this is the stuff it has `gathered` as the result. I guess something like `output` would be a better name.

Reply:
Done.